src/sql/workbench/parts/notebook/common/models/notebookTextFileModel.ts

```diff
@@ -0,0 +1,256 @@
+		let cellGuidRange = this.getCellNodeByGuid(textEditorModel, contentChange.cells[0].cellGuid);
+
+		// convert the range to leverage offsets in the json
+		if (contentChange && contentChange.modelContentChangedEvent && this.areRangePropertiesPopulated(cellGuidRange)) {
+			contentChange.modelContentChangedEvent.changes.forEach(change => {
+				let convertedRange: IRange = {
+					startLineNumber: change.range.startLineNumber + cellGuidRange.startLineNumber - 1,
+					endLineNumber: change.range.endLineNumber + cellGuidRange.startLineNumber - 1,
+					startColumn: change.range.startColumn + cellGuidRange.startColumn,
+					endColumn: change.range.endColumn + cellGuidRange.startColumn
+				};
+				// Need to subtract one because we're going from 1-based to 0-based
+				let startSpaces: string = ' '.repeat(cellGuidRange.startColumn - 1);
+				// The text here transforms a string from 'This is a string\n this is another string' to:
+				//     This is a string
+				//     this is another string
+				textEditorModel.textEditorModel.applyEdits([{
+					range: new Range(convertedRange.startLineNumber, convertedRange.startColumn, convertedRange.endLineNumber, convertedRange.endColumn),
+					text: change.text.split('\n').join('\\n\",\n'.concat(startSpaces).concat('\"'))
+				}]);
+			});
+		} else {
+			return false;
+		}
+		return true;
+	}
+
+	public transformAndApplyEditForOutputUpdate(contentChange: NotebookContentChange, textEditorModel: TextFileEditorModel | UntitledEditorModel): boolean {
+		if (Array.isArray(contentChange.cells[0].outputs) && contentChange.cells[0].outputs.length > 0) {
+			let newOutput = JSON.stringify(contentChange.cells[0].outputs[contentChange.cells[0].outputs.length - 1], undefined, '    ');
+			if (contentChange.cells[0].outputs.length > 1) {
+				newOutput = ', '.concat(newOutput);
+			} else {
+				newOutput = '\n'.concat(newOutput).concat('\n');
+			}
+			let range = this.getEndOfOutputs(textEditorModel, contentChange.cells[0].cellGuid);
+			if (range) {
+				textEditorModel.textEditorModel.applyEdits([{
+					range: new Range(range.startLineNumber, range.startColumn, range.startLineNumber, range.startColumn),
+					text: newOutput
+				}]);
+			}
+		} else {
+			return false;
+		}
+		return true;
+	}
+
+	public transformAndApplyEditForCellUpdated(contentChange: NotebookContentChange, textEditorModel: TextFileEditorModel | UntitledEditorModel): boolean {
+		let executionCountMatch = this.getExecutionCountRange(textEditorModel, contentChange.cells[0].cellGuid);
+		if (executionCountMatch && executionCountMatch.range) {
+			// Execution count can be between 0 and n characters long
+			let beginExecutionCountColumn = executionCountMatch.range.endColumn;
+			let endExecutionCountColumn = beginExecutionCountColumn + 1;
+			while (textEditorModel.textEditorModel.getLineContent(executionCountMatch.range.endLineNumber)[endExecutionCountColumn - 1]) {
```

Comment:
>textEditorModel.textEditorModel.getLineContent(executionCountMatch.range.endLineNumber) [](start = 10, length = 87)

Instead of calling getLineContent each iteration why not just call it once and then iterate over that?

Reply:
Awesome, thanks!